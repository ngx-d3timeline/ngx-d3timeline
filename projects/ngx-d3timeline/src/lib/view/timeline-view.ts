export class TimelineView {
  readonly width: number;
  readonly height: number;
  readonly margin: number;

  readonly bottom: number;
  readonly left: number;
  readonly right: number;
  readonly top: number;

  constructor({
    width,
    height,
    margin
  }: {
    width: number;
    height: number;
    margin: number;
  }) {
    this.width = width;
    this.height = height;
    this.margin = margin;

    this.bottom = height - margin;
    this.left = 0;
    this.right = width - margin;
    this.top = 0;
  }
}
