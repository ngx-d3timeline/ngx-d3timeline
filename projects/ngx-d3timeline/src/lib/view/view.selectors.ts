import { selectView } from '../store/state';
import {
  getViewTopLeft,
  getViewClipRectangle,
  getViewClipRectangleHeight,
  getViewClipRectangleWidth
} from './view-utils';
import { createSelector } from '../store-lib/selector/create-selector';

export const selectViewTopLeft = createSelector(selectView, getViewTopLeft);

export const selectViewClipRectangleHeight = createSelector(
  selectView,
  getViewClipRectangleHeight
);

export const selectViewClipRectangleWidth = createSelector(
  selectView,
  getViewClipRectangleWidth
);

export const selectViewClipRectangle = createSelector(
  selectViewTopLeft,
  selectViewClipRectangleWidth,
  selectViewClipRectangleHeight,
  getViewClipRectangle
);
