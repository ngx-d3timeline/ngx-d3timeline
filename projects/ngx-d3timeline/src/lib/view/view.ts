export class View {
  readonly width: number;
  readonly height: number;
  readonly margin = 50;

  readonly left = this.margin;
  readonly top = this.margin;
  readonly right: number;
  readonly bottom: number;

  constructor([width, height]: [number, number]) {
    this.width = width;
    this.height = height;

    this.bottom = height - this.margin;
    this.right = width - this.margin;
  }
}
