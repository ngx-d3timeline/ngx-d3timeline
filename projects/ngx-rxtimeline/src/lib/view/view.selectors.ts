import { selectView } from '../store/state';

import { createSelector } from '../store-lib/selector/create-selector';
import { Point } from '../core/point';
import {
  createStructuredSelector,
  createEnumSelector
} from '../store-lib/selector/selector-utils';
import { Rectangle } from '../core/rectangle';
import { toArray, sum } from '../core/array-utils';
import { pipe, clampZero, subtract, double } from '../core/function-utils';
import { constSelector } from '../store-lib/selector/selector';
import { Orientation } from '../core/orientation';
import {
  selectTimeOrientation,
  selectAxisOrientation
} from '../options/selectors/options.selectors';
import { AxisType } from '../axis/axis';

export const selectVerticalMargin = constSelector(50);
export const selectMarginRight = constSelector(50);
const selectMarginLeftWhenHorizontal = constSelector(100);
const selectMarginLeftWhenVertical = constSelector(50);

export const selectMarginLeft = createEnumSelector<Orientation, number>({
  Horizontal: selectMarginLeftWhenHorizontal,
  Vertical: selectMarginLeftWhenVertical
})(selectTimeOrientation);

const selectVerticalMarginTotal = createSelector(selectVerticalMargin, double);
const selectHorizontalMarginTotal = createSelector(
  selectMarginLeft,
  selectMarginRight,
  sum
);

const selectViewWidth = createSelector(selectView, view => view && view.width);
const selectViewHeight = createSelector(
  selectView,
  view => view && view.height
);

const selectViewLeft = createSelector(
  selectViewWidth,
  selectMarginLeft,
  Math.min
);

const selectViewTop = createSelector(
  selectViewHeight,
  selectVerticalMargin,
  Math.min
);

const selectViewRight = createSelector(
  selectViewWidth,
  selectMarginRight,
  subtract
);

const selectViewBottom = createSelector(
  selectViewHeight,
  selectVerticalMargin,
  subtract
);

export const selectViewTopLeft = createStructuredSelector<Point>({
  x: selectViewLeft,
  y: selectViewTop
});

export const selectViewClipRectHeight = createSelector(
  selectViewHeight,
  selectVerticalMarginTotal,
  pipe(subtract, clampZero)
);

export const selectViewClipRectWidth = createSelector(
  selectViewWidth,
  selectHorizontalMarginTotal,
  pipe(subtract, clampZero)
);

export const selectViewClipRect = createStructuredSelector<Rectangle>({
  height: selectViewClipRectHeight,
  width: selectViewClipRectWidth,
  x: selectViewLeft,
  y: selectViewTop
});

export const selectViewVerticalRange = createSelector(
  selectViewTop,
  selectViewBottom,
  toArray
);

export const selectViewHorizontalRange = createSelector(
  selectViewLeft,
  selectViewRight,
  toArray
);

export const selectViewRange = (axisType: AxisType) =>
  createEnumSelector<Orientation, number[]>({
    Vertical: selectViewVerticalRange,
    Horizontal: selectViewHorizontalRange
  })(selectAxisOrientation(axisType));
