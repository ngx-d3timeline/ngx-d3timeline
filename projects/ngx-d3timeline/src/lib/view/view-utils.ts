import { View } from './view';
import { Point } from '../core/point';
import { Rectangle } from '../core/rectangle';

export function getViewTopLeft(view: View): Point {
  return { x: view.left, y: view.top };
}

export function getViewClipRectangle(view: View): Rectangle {
  return {
    width: view.width - 2 * view.margin,
    height: view.height - 2 * view.margin,
    x: view.left,
    y: view.top
  };
}
