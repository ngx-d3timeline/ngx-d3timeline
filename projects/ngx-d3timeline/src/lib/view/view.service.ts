import { Injectable } from '@angular/core';
import { TimelineView } from './timeline-view';
import { BehaviorSubject } from 'rxjs';
import { map, tap } from 'rxjs/operators';

@Injectable({ providedIn: 'root' })
export class ViewService {
  readonly margin = 50;

  private viewSubject = new BehaviorSubject<[number, number]>(null);
  view$ = this.viewSubject.pipe(
    map(([width, height]) => new TimelineView(width, height, this.margin))
  );

  get rootTransform() {
    return `translate(${this.margin}, ${this.margin})`;
  }

  setView([width, height]: [number, number]) {
    this.viewSubject.next([width, height]);
  }
}
