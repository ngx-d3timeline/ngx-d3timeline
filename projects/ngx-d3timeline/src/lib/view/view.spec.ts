import { View } from './view';

describe('View', () => {
  it('should return correct bounds', () => {
    const view = new View([300, 400]);

    expect(view).toMatchSnapshot();
  });
});
