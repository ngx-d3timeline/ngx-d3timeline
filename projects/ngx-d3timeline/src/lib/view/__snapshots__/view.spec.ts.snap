// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View should return correct bounds 1`] = `
View {
  "bottom": 350,
  "height": 400,
  "left": 50,
  "margin": 50,
  "right": 250,
  "top": 50,
  "width": 300,
}
`;
